Progress Report May 2022 (#323)
yuzu-emu/yuzu-emu.github.io: site/content/entry/yuzu-progress-report-may-2022/index.md

+++
date = "2022-06-10T05:00:00-03:00"
title = "Progress Report May 2022"
author = "GoldenX86"
coauthor = ""
forum = 0
+++

> Greetings yuz-ers. While this monthly report will not be as extensive as others, we have several critical changes to discuss. Roll the tape!

Reviewer: Remove " not as extensive" and instead say something like:

"This time around, we're covering many small and incremental improvements to yuzu. Rest assured, we also have some *major* rewrites and improvements in the works, and we'll touch on those near the end. Roll the tape!"

In the meantime, toastUnlimited {{< gh-hovercard "8379" "blocked the extension" >}} on AMD drivers with affected Vulkan driver versions.

[asLody](https://github.com/asLody) {{< gh-hovercard "8311" "implemented stencil fixes when two faces are disabled." >}}
This has the potential to improve rendering in native OpenGL games.

## HLE Improvements

In software engineering, a spinlock is a lock that causes a thread trying to acquire it to simply wait in a loop
 ("spin") while repeatedly checking whether the lock is available.

{{< imgs
    "./spinlock.png| Example of a spinlock"
  >}}

On the other hand, the word "mutex" stands for an object providing `MUTual EXclusion` between threads. 
Mutex ensures that only one thread has access to a critical section or data by using operations like a lock and unlock. 
A thread that acquires the lock of mutex can use the critical section while other threads must wait till the lock is released.

{{< imgs
    "./mutex.png| Example of a mutex"
  >}}

In theory, when a thread tries to lock a mutex and it does not succeed, because the mutex is already locked, it will go to sleep, immediately allowing another thread to run. 

> It will continue to sleep until being woken up, which will be the case once the mutex is being unlocked by whatever thread was holding the lock before.

Reviewer: The paused thread will continue to "sleep" until it is able to acquire the mutex. This may happen once the current thread holding the mutex lock releases it.

But when a thread tries to lock a spinlock and it does not succeed, it will continuously re-try locking it, until it finally succeeds; thus it will not allow another thread to take its place until the OS forces it to timeout.

> Hence, polling on a spinlock will constantly waste CPU time and if the lock is held for a longer amount of time, this will waste a lot more CPU time and it would have been much better if the thread was sleeping instead.

Reviewer: Remove this sentence and replace with:

Consequently, threads "spinning" to acquire the lock will waste (perhaps precious) system resources. While the Switch's own operating system uses spinlocks, this drain on resources can be problematic when emulating on lower-end hardware. Using the host operating system (e.g. Windows or Linux) mutex allows yuzu to continue emulation tasks on other available threads.

> However in practice, most modern operating systems (Windows/Linux) use hybrid mutexes and hybrid spinlocks.

Reviewer: Helpfully, (instead of However in practice)

> So spinlocks are fine on systems with tons of cores, as it's not a big deal for host threads to spin. 

Reviewer: As a note, the spinlock approach works fine with systems with threads to spare. Furthermore, this approach would also be quite viable if yuzu were able to emulate with exactly 4 threads, as the Switch has 4 cores. However, for emulation, we need many more threads (e.g. for UI, audio, GPU emulation, etc.), so this approach is not quite practical at this time.

> On low core counts, you really start starving the system by doing that, whereas a mutex will allow another thread to be scheduled on that core/thread.

Reviewer: perhaps remove with the above explanation

Thus {{< gh-hovercard "8172" "by moving from spinlocks to mutexes," >}} we were able to improve how yuzu runs on systems with low core counts. 

> Our testing results showed that yuzu is now much more usable on 4 thread systems.

Reviewer: Could use some more anecdotes here on what this improves, since it was kind of a big deal!

## UI changes

> [Docteh](https://github.com/Docteh) has been very helpful with some translation holes we had for a while.

Reviewer: Could use another segway as usual... e.g.

"Pivoting towards user interface improvements, a newer contributor to the team has stepped up with some great changes!" (or something)

For example, the Custom RTC setting had several issues if the Windows system locale was set in certain languages, making it either display incorrectly (for example lacking the AM/PM indicator), or completely unusable.
{{< gh-hovercard "8291" "Fixing the display format" >}} allows Custom RTC to show up correctly in any language now.

The Network tab in `Emulation > Configure… > System` could remain untranslated after changing languages. This was a simple case of forgetting to include the tab in the translations, so Docteh {{< gh-hovercard "8293" "fixed the oopsie" >}} and the lone Network tab now displays as it should.

{{< single-title-imgs-compare
    "Netto-kun. Now I want a Battle Network Legacy Collection, c'mon CAPCOM!"
    "./netbug.png"
    "./netfix.png"
>}}

> yuzu’s icon in the About dialog has been in a struggle lately, a fix for Linux builds would break Windows support, and vice versa.

Reviewer: "been in a struggle" -> replace with more precise language, e.g. "For a while now, the layout of yuzu's About dialog, particularly on Linux, has had some issues. While we've attempted to fix it in the past, these attempts would have adverse effects on other platforms..." blah blah blah

Via qtcreator, Docteh {{< gh-hovercard "8339" "fixed the About dialog UI file," >}} and removed an old warning caused by the original .png image.

> Thanks for taking the time to properly address the issue once and for all!

Reviewer: Thank you Docteh ... !

(otherwise it reads as you are thanking the reader)

## Controller changes

> Motion continued reporting data when disabled, causing `Pokémon Let’s Go, Eevee/Pikachu!` to spam `StopSixAxisSensor` errors in the log.

Reviewer: Segway segway segway -- did Flam even ask for this before? <3

"Onward to the land of german77, where he continues his unrelenting quest to deliver the best user input experience possible in **any emulator ever**!"

While working on this, [german77](https://github.com/german77) noticed a missing parameter, `delta_time`.
Its implementation allows yuzu to have an {{< gh-hovercard "8308" "accurate motion refresh rate," >}} equal to the Switch.

In an all-in-one pull request, german77 made {{< gh-hovercard "8368" "several input changes," >}} including:

- Add proper error handling for several [HID](https://en.wikipedia.org/wiki/Human_interface_device) functions
- Improve previous implementations to match more closely to native hardware.
- Implement functions needed by `Nintendo Switch Sports`, `EnableSixAxisSensorUnalteredPassthrough`, `IsSixAxisSensorUnalteredPassthroughEnabled`, `LoadSixAxisSensorCalibrationParameter`, `GetSixAxisSensorIcInformation`, `ResetIsSixAxisSensorDeviceNewlyAssigned`.

> Nintendo Switch Sports will only be playable after an audio rewrite and more GPU work.

Reviewer: Replace with:

"While we've made some great progress here, Nintendo Switch Sports remains unplayable on yuzu until we rework our audio implementation, and furthermore make some much needed GPU fixes. While audio and perfect rendering may not seem critical to playability, games often are quite unstable if these are not accurate. Rest assured, we're working on these and will have more to share very soon!"

`Arcaea` was reported as having issues with touch emulation.

> Turns out this game checks the position on release and some input drivers didn't keep it's position on release. 

Reviewer: * touch release
* it's == it is, engrish pls

> Also, multi-touch wasn't working on touch screens.

Reviewer: Additionally, ...

After performing basically {{< gh-hovercard "8372" "a mini-rewrite of the touch emulation," >}} german77 fixed both issues.

{{< imgs
    "./arcaea.png| Osu! but better? (Arcaea)"
  >}}

> One of the hurdles when working with an infinite amount of different controllers is the different quality of implementations they have.

Reviewer: infinite -> near endless

Since yuzu used to wait for the controller to respond after sending a vibration signal, slow controllers could stall the whole emulator, causing severe stuttering.

> To counter this, german77 {{< gh-hovercard "8374" "moved vibration to a queue in a separate thread," >}} allowing yuzu to move along with emulation while our controllers try to do their best.

Reviewer: ... allowing yuzu to move along with emulation, letting your controller make its best effort.

> An example of why more CPU threads help more than few faster threads, there’s always some background process that could be made asynchronous.

Reviewer: This sentence is hard to understand. Perhaps:

"This is just another example of how emulation can often be improved by moving blocking operations to asynchronous background threads to improve overall usability. In fact, yuzu uses dozens of threads for emulation, which is all the more reason why eliminating spinlocks really helps things to run as smooth as butter!"

(bet you liked that tie in to the section above, huh?)

## Future projects

> [Blinkhawk](https://github.com/FernandoS27) has been waiting for NVflinger to be back on track before continuing with `Project Y.F.C.`, so work can resume now.

Reviewer: Reflect the present -- "While YFC was slightly stalled due to some NVFlinger regressions, these have since been resolved and will be covered in the next Progress Report! YFC is making great progress and is on track to release soon. As a reminder, YFC is an overhaul of various parts of our GPU emulation, fixing many inaccuracies and improving performance."

(or something like that, we should stop assuming every reader knows our silly project names)

> [Maide](https://github.com/Kelebek1) is up to something.

Reviewer: Again, add a little more, e.g. 

"... As a hint, Maide has made some great audio improvements in the past :) "

And toastUnlimited is working on getting MinGW Clang builds for Windows, which could potentially be faster than the MSVC builds we’re using now. This work is tied with the release of `Project Gaia`, so it will take a bit.

## Bonus track

As an extra bonus, gidoly, one of our team members recently got his hands on a Ryzen 5800X3D, giving us the chance to compare it to a regular 5800X fixed at 4.5GHz so only the extra cache should be relevant.

Here are the results!

{{< imgs
    "./chart.png| Still the best upgrade path for a Zen1 user"
  >}}

While the 5800X is manually forced to a 4.5GHz frequency, the 5800X3D naturally tops out at a 4.45GHz clock speed, the results are respectable, but nothing amazing.
Pokémon Brilliant Diamond certainly loves the extra cache, while Metroid Dread is punished by its extra latency.

> That’s all folks! As always, a pleasure to have you here. See you next month!

Reviewer: Add another sentence, this is a weak sauce ending.

> That’s all folks! As always, thank you for your support, and we hope that you enjoyed this summary of our recent progress. See you next month! Until then, keep on emulating, and let us know what we can do to make yuzu the best possible emulation experience!

> Some content had to be pushed to the next progress report, we're sorry for the inconvenience.

Reviewer: Meh, remove this last line

+++
date = "2022-06-12T17:50:00-03:00"
title = "Progress Report May 2022"
author = "GoldenX86"
coauthor = "CaptV0rt3x"
forum = 585658
+++

Greetings yuz-ers. This time around, we're covering small and incremental improvements to yuzu. Rest assured, we also have some major rewrites and improvements in the works, and we'll touch on those near the end. Roll the tape!

<!--more--> 

## End of support for EOL Windows versions

Let’s first address the elephant in the room, shall we?

While working on dynarmic and kernel emulation, including improving the compatibility of 4 thread CPU systems, we made changes to [dynarmic](https://github.com/merryhime/dynarmic) and [fastmem](https://yuzu-emu.org/entry/yuzu-fastmem/) that broke support for Windows 10 revision 1803 and older, including Windows 7 and Windows 8/8.1.

> While it was a purely accidental change, this was notified with the introduction of fastmem.

Reviewer: Replace with:

> While fastmem was only ever designed to work with newer operating systems, the changes to dynarmic breaking support for older Windows versions was purely accidental. That being said, it is yet another sign of the times, and that a pre-Windows 10 experience in yuzu will continue to become more subpar.

Finally, projects like [Dolphin](https://www.reddit.com/r/emulation/comments/utbpmm/dolphin_has_dropped_support_for_windows_7_and_8/) have already followed the same path, and for the same exact reasons.

A 13 years old Windows is old enough to drive in some places.

For those that still prefer to not upgrade, [Mainline 990](https://github.com/yuzu-emu/yuzu-mainline/releases/download/mainline-0-990/yuzu-windows-msvc-20220419-a5e7c5330.zip) and older will work just fine.

## Vulkan by default

[As previously discussed](https://yuzu-emu.org/entry/yuzu-progress-report-feb-2022/#vulkan-is-the-future), we have to circumvent issues like OEM-locked drivers (so common on Intel hardware, [it has its own official procedure](https://www.intel.com/content/www/us/en/support/articles/000056629/graphics.html)) and broken third party software limitations (outdated screen recorders are a common cause of broken rendering) in order to provide a smooth experience with Vulkan as the default API.

The two main causes for Vulkan related crashes when trying to boot a game or opening yuzu’s configuration are:

> - Broken Vulkan layers on HUD and screen recording software, could cause issues when yuzu and drivers add support for new Vulkan extensions. Keeping software up to date is the only way to prevent this issue from happening.

Reviewer: Remove comma

{{< imgs
    "./ok.png| Check passes, Vulkan works!"
  >}}

If this check fails, a warning will be displayed the next time you launch yuzu. If this happens, you will only be able to use OpenGL as the graphics API. You will still have the option to pick the shader backend (GLSL, GLASM, SPIR-V) that best suits your needs.

{{< imgs
    "./error.png| Oh oh.."
  >}}

For those that happen to land in this situation, a button labeled "Check for Working Vulkan" at the bottom of the Graphics settings window will show up, allowing to retest Vulkan support.

{{< imgs
    "./button.png| Once you manage to solve the issue, click on the button at the bottom!"
  >}}

Thanks to [toastUnlimited](https://github.com/lat9nq), gone is OpenGL as the default graphics API. 
Out with the old, in with the new. {{< gh-hovercard "8393" "Long live King Vulkan." >}}

Going forward, Vulkan will be the top priority for our developers, but they will still continue to support OpenGL.
OpenGL users are recommended to use the GLSL shader backend, as GLASM and SPIR-V will receive limited support from now on.

## Graphical changes, driver issues, and the nostalgia bliss that is the good old 64

> This month, [byte[]](https://github.com/liamwhite) continues the wave of improvements for `Super Mario 3D All-Stars`, this time, he noticed a bug in the DMAcopy ([direct memory access](https://en.wikipedia.org/wiki/Direct_memory_access)) of the Nintendo Switch’s GPU.

Reviewer: This past month, ... continued ...

Reviewer: Period after Super Mario 3D All-Stars. This time, he notied...

While this wasn't an issue while using Vulkan (performance aside), now `Super Mario 64` and `Super Mario Galaxy` {{< gh-hovercard "8314" "are playable in both graphics APIs." >}} 
Fermi GPU users rejoyce.

{{< imgs
    "./sm64.png| It's-a Mario, now running fast in OpenGL! (Super Mario 64)"
  >}}

One of the important parts of yuzu's graphical emulation is the need to translate small sets of GPU instructions, called `macros`. 
yuzu uses a Just-in-Time (JIT) compiler to execute these macros in a performant way. It provides a performance boost of about 10% over interpretation in most cases.

byte[] found that due to emulation inaccuracies, sometimes a macro could try to access a parameter that was too far outside the bounds of what it was supposed to be accessing. This could crash the emulator without a single trace as to why in some cases.
{{< gh-hovercard "8319" "One less reason for annoying crashes." >}}

Additionally, byte[] added the option to {{< gh-hovercard "8320" "dump all macros" >}} used by a game for debugging purposes.

But why are macros important enough to merit their own dump mechanism?

> Turns out, the `Nintendo 64` emulator (*totally not outside Nintendo's Terms of Service*), included with the `Nintendo Switch Online` subscription, reassigns the same macros multiple times, each time with different code. yuzu incorrectly appended the new code to the end of the macro in this case, instead of replacing the existing code.

Reviewer: Nintendo Switch Online (NSO)

(always define acronyms you later use)

This improves emulation performance with literally every game, and to a varying degree, on any CPU.
Let's dive in.

In software engineering, a [spinlock](https://en.wikipedia.org/wiki/Spinlock) is a lock that causes a thread trying to acquire it to simply wait in a loop
 ("spin") while repeatedly checking whether the lock is available.

{{< imgs
    "./spinlock.png| Example of a spinlock, simple but gets the job done"
  >}}

There exists another synchronization primitive with a similar function, [the mutex](https://en.wikipedia.org/wiki/Mutual_exclusion). 

The word "mutex" stands for an object providing `MUTual EXclusion` between threads.
A mutex ensures that only one thread has access to a critical section or data by using operations like a lock and unlock.
A critical section is a shared resource that many threads want to access.
While there is no issue if multiple threads want to read the same critical section, no new thread can modify the section until the previous thread finishes its own writing. 
Under this scenario, the first thread locks the section, and will remain that way until the lock is released.

{{< imgs
    "./mutex.png| Example of a mutex"
  >}}

> In theory, when a thread tries to lock a mutex and it does not succeed (for example because the mutex is already locked), it will be `paused`.

Reviewer: I would remove the highlight of paused and sleep; This is not necessarily official terminology, and more of just  a way to explain it in a not overly technical way

The operating system will then take the opportunity to schedule an available and ready thread to run in its place.
The paused thread will continue to `sleep` until it is able to acquire the mutex. 
This may happen once the current thread holding the mutex lock releases it.

Consequently, threads "spinning" to acquire the lock will waste (perhaps precious) system resources. 
While the Switch's own operating system uses spinlocks, this drain on resources can be problematic when emulating on lower-end hardware. 
Using the host operating system (Windows or Linux) mutex allows yuzu to continue emulation tasks on other available threads.

Helpfully, most modern operating systems use hybrid mutexes and hybrid spinlocks.
The spinlock approach would work fine on systems with threads to spare. 

> However, for emulation, we need many threads (for UI, audio, GPU emulation, logging, etc.), so this approach is not quite practical, especially on CPUs with low core/thread counts.

Reviewer: Replace practical with ideal

OpenGL users are recommended to use the GLSL shader backend, as GLASM and SPIR-V will receive limited support from now on.

## Graphical changes, driver issues, and the nostalgia bliss that is the good old 64

This month, [byte[]](https://github.com/liamwhite) continues the wave of improvements for `Super Mario 3D All-Stars`, this time, he noticed a bug in the DMAcopy ([direct memory access](https://en.wikipedia.org/wiki/Direct_memory_access)) of the Nintendo Switch’s GPU.

`DMACopy` is a mechanism that many games use to send texture data to the GPU, it handles the format conversion from "pitch" (pixels on a line by line basis) to "tiled" (gridded) images.
This process works by writing the pitch image data into GPU memory accessible by the DMA engine. Next, a DMAcopy is requested through the DMA engine driver, converting the image data into a separate buffer accessible by the GPU. This buffer will then be used as the texture on the final draw.

{{< gh-hovercard "8313" "After fixing `bytes_per_pixel`," >}} `Super Mario Galaxy` now has proper lens flare.

{{< single-title-imgs-compare
    "RTX On? (Super Mario Galaxy)"
    "./dmabug.png"
    "./dmafix.png"
>}}

> byte[] also improved the way OpenGL interprets face flips depth, [replacing the previously reported fix](https://yuzu-emu.org/entry/yuzu-progress-report-apr-2022/#saving-princess-peach-yet-again). The previous implementation had terrible peformance in OpenGL, making it virtually unplayable.

Reviewer 2: The current implementation has poor performance in _Vulkan_. OpenGL runs at full speed, which was the motivation to fix the rendering in OpenGL.

Reviewer 2: Also the me getting face flips wrong the first time is just because it was tricky to test due to being an extremely uncommon configuration for the GPU.

While this wasn't an issue while using Vulkan, now `Super Mario 64` and `Super Mario Galaxy` {{< gh-hovercard "8314" "are playable in both APIs." >}} 
Fermi GPU users rejoyce.

{{< imgs
    "./sm64.png| It's-a Mario, now running fast in OpenGL! (Super Mario 64)"
  >}}

> One of the important parts of yuzu's graphical emulation is the need to translate small sets of GPU instructions, called `macros`. 
> yuzu uses a Just-in-Time (JIT) compiler to emulate these macros in a performant way.
> MacroJIT is an optimization applied over this process, it manages to provide a performance boost of plus 10%, more or less.

Reviewer 2: ```suggestion
One of the important parts of yuzu's graphical emulation is the need to translate small sets of GPU instructions, called `macros`. 
yuzu uses a Just-in-Time (JIT) compiler to execute these macros in a performant way. It provides a performance boost of about 10% over interpretation in most cases.
```

> byte[] found that macro could try to access a parameter that was too far outside the bounds of what it was supposed to be accessing. 
> Usually the end result of this behaviour would be the emulator crashing.
> {{< gh-hovercard "8319" "One less reason for annoying crashes." >}}

Reviewer 2: ```suggestion
byte[] found that due to emulation inaccuracies, sometimes a macro could try to access a parameter that was too far outside the bounds of what it was supposed to be accessing. This could crash the emulator without a single trace as to why in some cases.
{{< gh-hovercard "8319" "One less reason for annoying crashes." >}}
```

> Now, while on the subject or macros, MME, or Macro Method Expander, are small programs sent to the GPU when booting a game, responsible for executing methods (anything that changes the current status of the GPU).
> byte[] added the option to {{< gh-hovercard "8320" "dump such MME macros" >}} for debugging purposes.

Reviewer 2: MME is just the macro engine
```suggestion
Additionally, byte[] added the option to {{< gh-hovercard "8320" "dump all macros" >}} used by a game for debugging purposes.
```

> But why are MME macros important enough to merit their own dump mechanism?

Reviewer 2: ```suggestion
But why are macros important enough to merit their own dump mechanism?
```

> Turns out, the `Nintendo 64` emulator (*totally not outside Nintendo's Terms of Service*), included with the `Nintendo Switch Online` subscription, reasigns the same macros multiple times, each with different code.

Reviewer 2: ```suggestion
Turns out, the `Nintendo 64` emulator (*totally not outside Nintendo's Terms of Service*), included with the `Nintendo Switch Online` subscription, reassigns the same macros multiple times, each time with different code. yuzu incorrectly appended the new code to the end of the macro in this case, instead of replacing the existing code.
```

forum = 0
+++

Greetings yuz-ers. This time around, we're covering small and incremental improvements to yuzu. Rest assured, we also have some major rewrites and improvements in the works, and we'll touch on those near the end. Roll the tape!

<!--more--> 

## End of support for EOL Windows versions

Let’s first address the elephant in the room, shall we?

While working on dynarmic and kernel emulation, including improving the compatibility of 4 thread CPU systems, we made changes to [dynarmic](https://github.com/merryhime/dynarmic) and [fastmem](https://yuzu-emu.org/entry/yuzu-fastmem/) that broke support for Windows 10 revision 1803 and older, including Windows 7 and Windows 8/8.1.

> While it was a purely accidental change, this was warned with the introduction of fastmem.

Reviewer 3: Maybe use "notified" or "announced" or "informed" or "made clear" instead of warned?

Due to our focus on improving accuracy, stability and performance, it doesn't make much sense to divert time and resources onto maintaining old and out of support operating systems. 
From Mainline version 991 and onward, only Windows 10 revision 1809 and newer, Windows 11, and Linux, will be the officially supported operating systems.

This decision is reinforced by the lack of GPU driver support on EOL systems (which affects Vulkan support going forward), inconsistencies in long file path limitations (critical for file system emulation improvements), as well as worse memory handling on a kernel level, which is required to properly emulate the Switch and its subsystems.

Without forcing the developers to divert time in supporting dated platforms (which themselves no longer use), they can instead focus their energy on improving the core components of Nintendo Switch Emulation.

Finally, projects like [Dolphin](https://www.reddit.com/r/emulation/comments/utbpmm/dolphin_has_dropped_support_for_windows_7_and_8/) have already followed the same path, and for the same exact reasons.

> A 13 years old Windows is old enough to drive in some places.

Reviewer 3: Necessary?

For those that still prefer to not upgrade, [Mainline 990](https://github.com/yuzu-emu/yuzu-mainline/releases/download/mainline-0-990/yuzu-windows-msvc-20220419-a5e7c5330.zip) and older will work just fine.

## Vulkan by default

> [As previously discussed](https://yuzu-emu.org/entry/yuzu-progress-report-feb-2022/#vulkan-is-the-future), we have to circumvent issues like vendor-locking drivers (so common on Intel hardware, [it has its own official procedure](https://www.intel.com/content/www/us/en/support/articles/000056629/graphics.html)) and broken third party software limitations (outdated screen recorders are a common cause of broken rendering) in order to provide a smooth experience with Vulkan as the default API.

Reviewer 3: Vendor-locking drivers?

> The two main reasons for Vulkan related crashes when trying to boot a game or opening yuzu’s configuration are:

Reviewer 3: main causes

> - Broken Vulkan layers on HUD and screen recording software, which could cause issues when yuzu and drivers add support for new Vulkan extensions. Keeping software up to date is the only way to prevent this issue from happening.

Reviewer 3: ```diff
- software, which could ...
+ software, could ...
```

    "./ok.png| Check passes, Vulkan works!"
  >}}

> If this check fails, a warning will be displayed at the next time you launch yuzu. If this happens, you will only be able to use OpenGL as the graphics API. You will still have the option to pick the shader backend (GLSL, GLASM, SPIR-V) that best suits your needs.

Reviewer 3: ```diff
- displayed at the next time...
+ displayed the next time...
```

{{< imgs
    "./error.png| Oh oh.."
  >}}

> For those souls that happen to land in this situation, a button labeled "Check for Working Vulkan" at the bottom of the Graphics settings window will show up, allowing to retest Vulkan support.

Reviewer 3: ```diff 
- souls
```

## Graphical changes, driver issues, and the nostalgia bliss that is the good old 64

> This pont month, [byte[]](https://github.com/liamwhite) continued the wave of improvements for `Super Mario 3D All-Stars`.

Reviewer: pont

Since yuzu used to wait for the controller to respond after sending a vibration signal, slow controllers could stall the whole emulator, causing severe stuttering.
To counter this, german77 {{< gh-hovercard "8374" "moved vibration to a queue in a separate thread," >}} allowing yuzu to move along with emulation, letting your controller make its best effort.
This is just another example of how emulation can often be improved by moving blocking operations to asynchronous background threads to improve overall usability. 
In fact, yuzu uses dozens of threads for emulation, which is all the more reason why eliminating spinlocks really helps things to run as smooth as butter!

## Future projects

While `Project Y.F.C.` was slightly stalled due to some NVFlinger regressions, these have since been resolved and will be covered in the next progress report!
Under [blinkhawk](https://github.com/FernandoS27)'s lead, `Project Y.F.C.` is making great progress and is on track to release soon.
As a reminder, `Project Y.F.C.` is an overhaul of various parts of our GPU emulation, fixing many inaccuracies and improving both performance and compatibility.

> [Maide](https://github.com/Kelebek1) is up to something. Let me provide a hint, if you check previous progress reports, you will notice a common theme with Maide's pull requests.

Reviewer 3: >  {... } something. (Hint: if you check the previous progress reports, you'll notice a common theme with their pull requests)
